logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -225,9 +225,9 @@ private void openPages() throws IOException {
             }
             final Checkpoint cp = this.checkpointIO.read(cpFileName);
 
-            logger.debug("opening tail page: {}, in: {}, with checkpoint: {}", pageNum, this.dirPath, cp.toString());
+            logger.debug("opening tail page: {}, in: {}, with checkpoint: {}", pageNum, this.dirPath, cp);
 
-            PageIO pageIO = new MmapPageIOV2(pageNum, this.pageCapacity, this.dirPath);
+            MmapPageIOV2 pageIO = new MmapPageIOV2(pageNum, this.pageCapacity, this.dirPath);
```

Comment:
hmm we could but do we need two expose it given it's only used for logging?
going to follow up and add it but I believe the concrete `MmapPageIOV2` type is only visible internally in the `Queue` class, the method `MmapPageIOV2` is passed to is private ...

Reply:
My concern on using the concrete class is that, if for example, the `pageIO` is added in added as a parameter in a public method invoked by this code, then we could have a potential publication of `MmapPageIOV2` to consumer classes. I think that `MmapPageIOV2` should be considered as a package private class, but it's public, because needs to be used by `PqRepair` in parent package `org.logstash.ackedqueue`.

I understand your point and resonate with me.

The log line has to print information on a page, what do you think if we leverage the `toString` method of the pageIO implementation?
I mean to override `toString`in both `MmapPageIOV2` and `MmapPageIOV1` and let the log to print it with:
```java
logger.debug(""tail page: {} does not exist", pageIO);
```
WDYT?

However I'm fine if you decide to keep the code as it is, I think the `toString` way is a good path, respect the design and respect the idea of the log line "provide the details to identify the page that hit the problem"